### DISCUSSION TITLE
How to implement DataTrigger in Avalonia

### DISCUSSION THREAD

**Discussion by Mochengvia** [AvaloniaUI/Avalonia#11486] [Q&A]
I want to convert a piece of WPF code to Avalonia.
If the property value of `HoverBackground` is not null, it is assigned to `(i:VisualStateHelper.Background)`.

```
<DataTrigger Binding="{Binding Path=(local:ButtonHelper.HoverBackground), Converter={StaticResource {x:Static rs:ConverterKeys.IsNonnullConverter}}, RelativeSource={RelativeSource Self}}"
             Value="True" />
    <Setter Property="(i:VisualStateHelper.Background)"
            Value="{Binding Path=(local:ButtonHelper.HoverBackground), RelativeSource={RelativeSource Self}}" />
</DataTrigger>
```

I try to use DataTriggerBehavior in the `Avalonia.Xaml.Behaviors` package to implement it, but I get an error: System.ArgumentException: "Could not find binding source: either target or anchor must be an IStyledElement."

The issue of the Avalonia.Xaml.Behaviors repo seems to be unresponsive, so I re-issue the question here.
Any suggestions? Or other similar implementation methods. Thanks.

```
<Style Selector="Button:pointerover">
  <Setter Property="(ia:Interaction.Behaviors)">
    <ia:BehaviorCollectionTemplate>
      <ia:BehaviorCollection>
        <iac:DataTriggerBehavior Binding="{Binding $parent[Button].(local:ButtonHelper.HoverBackground)}"
								 ComparisonCondition="NotEqual"
								 Value="{x:Null}">
          <iac:ChangePropertyAction TargetObject="$parent"
									PropertyName="(i:VisualStateHelper.Background)"
									Value="{Binding $parent[Button].(local:ButtonHelper.HoverBackground)}" />
        </iac:DataTriggerBehavior>
      </ia:BehaviorCollection>
    </ia:BehaviorCollectionTemplate>
  </Setter>
</Style>
```

**Comment by maxkatz6** [comment]
See comments and recommendations here https://github.com/AvaloniaUI/Avalonia/issues/1362

> The issue of the Avalonia.Xaml.Behaviors repo seems to be unresponsive, so I re-issue the question here.

I don't know why you have this issue.
But I also don't really understand what are you trying to do with your code.

> TargetObject="$parent"

Are you sure Behaviors support this syntax? It's not a binding...
Also "ChangePropertyAction" by itself probably doesn't have a parent, but I might be wrong and it uses style target.

**Comment by Mochengvia** [comment]
@maxkatz6 

We are trying to port PanuonUI to Avalonia. This is a WPF control library, developers can use attached properties to modify the style of hovering and clicking, instead of Style or Class.
For example, the developers can use this method to change the hover background of the button:
```
<Button Content="Button"
		Background="#FF0000"
		pu:ButtonHelper.HoverBackground="#00FF00" />
```

The value of `HoverBackground` property is user-set and mutable, so we have to use bindings instead of exact values.
If I use this :
```
<Style Selector="Button:pointerover">
	<Setter Property="Background" 
			Value="{Binding Path=(local:ButtonHelper.HoverBackground), RelativeSource={RelativeSource Self}}" />
</Style>
```

This causes a problem, if the value of `HoverBackground` is null (user did not specify a value), its background color will become transparent on hover. 
This is not the desired effect, so we must assign it to `Background` when the value of `HoverBackground` property is not null. In WPF, we can easily use DataTrigger to do this, is there any alternative in Avalonia?

I've looked at the conversation in #1362 before asking this question, but it's all about determining values rather than using bindings.

**Reply by timunie** [reply]
@Mochengvia I see some options (no ranking): 
1. Use a Converter 
2. Inside the getter of your property return the default value if `null`
3. Use a TargetNullValue in your Binding
4. Use Avalonia.Behaviors https://github.com/AvaloniaUI/Avalonia.Xaml.Behaviors

**Reply by Mochengvia** [reply]
Thank you for your reply.
In this case, we may have to consider using a method similar to DataTrigger, because when the property value of `HoverBackground` is null, we hope that the `Background` color will not change when the mouse hovers, but if we use `TargetNullValue` or the `defaultValue`, the control's background color will still change.
I tried using the `Avalonia.Xaml.Behavior` library (as the code I showed in the main body), but it didn't work as expected, and I couldn't find an example of binding to an attached property in the sample project, and the issue didn't respond either.
Please let me know if there are other possibilities.
